byterun/array.c

```diff
@@ -69,10 +69,31 @@ CAMLprim value caml_array_get_float(value array, value index)
 
 CAMLprim value caml_array_get(value array, value index)
 {
+#ifdef FLAT_FLOAT_ARRAY
   if (Tag_val(array) == Double_array_tag)
     return caml_array_get_float(array, index);
-  else
-    return caml_array_get_addr(array, index);
+#endif
+  CAMLassert (Tag_val(array) != Double_array_tag);
```

Comment:
What about putting the assertion under a #elseif to avoid it in legacy mode?  (But perhaps the C compilers are clever enough to avoid it.)

Reply:
What's the point? The assertion is also true in legacy mode, and more assertions are always better. Assertions are only compiled in the debug runtime anyway.